File: pkg/EarthSciAST.jl/test/classification_test.jl
# Derived variable classification (esm-spec §6.3.1) against the cross-binding
# oracle in `tests/conformance/classification/`.
#
# esm 1.0.0 declares two variable types and derives every finer category a
# solver needs. These fixtures are the one answer all five bindings are compared
# against, so this file asserts the goldens directly rather than re-stating the
# rules in Julia-local expectations.

using Test
using EarthSciAST
using JSON3

include("testutils.jl")

const _CLS_DIR = joinpath(TESTUTILS_REPO_ROOT, "tests", "conformance", "classification")
include(joinpath(@__DIR__, "..", "scripts", "classification_adapter.jl"))
const _CLS = ClassificationConformanceAdapter

@testset "Classification API (esm-spec §6.3.1)" begin
    manifest_path = joinpath(_CLS_DIR, "manifest.json")
    @test isfile(manifest_path)
    manifest = JSON3.read(read(manifest_path, String))

    @testset "golden agreement — $(String(fx["id"]))" for fx in manifest["fixtures"]
        fixture_path = joinpath(_CLS_DIR, String(fx["fixture"]))
        golden_path = joinpath(_CLS_DIR, String(fx["golden"]))
        @test isfile(fixture_path)
        @test isfile(golden_path)

        got = _CLS.classify_fixture(fixture_path)["models"]
        want = JSON3.read(read(golden_path, String))["models"]

        # Same set of model NODES: a binding that flattens the document before
        # classifying returns one merged node where the golden expects two.
        @test sort!(collect(keys(got))) == sort!(String[String(k) for k in keys(want)])

        for (mname, wmodel) in pairs(want)
            gmodel = got[String(mname)]
            for key in ("ode_states", "observed_unknowns", "algebraic_unknowns",
                        "brownian_parameters", "discrete_parameters",
                        "sampled_parameters", "constant_parameters")
                @test (String(mname), key, gmodel[key]) ==
                      (String(mname), key, String[String(x) for x in wmodel[key]])
            end
            @test (String(mname), gmodel["system_kind"]) ==
                  (String(mname), String(wmodel["system_kind"]))
            # `declared_system_kind` is absent from a golden that does not pin
            # it; where present, `null` means the model omits the field.
            if haskey(wmodel, "declared_system_kind")
                w = wmodel["declared_system_kind"]
                @test gmodel["declared_system_kind"] ==
                      (w === nothing ? nothing : String(w))
            end
        end
    end

    @testset "the sets PARTITION their side of the variables" begin
        # §6.3.1 states the partition normatively, so it is checked on every
        # fixture rather than trusted. `assert_classification_partitions` throws
        # on an overlap or a gap; `classify_fixture` calls it per model node.
        for fx in manifest["fixtures"]
            path = joinpath(_CLS_DIR, String(fx["fixture"]))
            @test _CLS.classify_fixture(path) isa AbstractDict
        end
    end

    @testset "is_ode_state agrees with ode_states" begin
        file = EarthSciAST.load(joinpath(_CLS_DIR, "fixtures", "basic_partition.esm"))
        m = file.models["M"]
        states = Set(ode_states(m))
        for name in keys(m.variables)
            @test is_ode_state(m, name) == (name in states)
        end
        # A name the model does not declare is not an ODE state, and asking is
        # not an error.
        @test is_ode_state(m, "not_declared") == false
    end

    @testset "a derivative LHS may be wrapped (D(u), D(u[i]), aggregate{D(...)})" begin
        # `tests/valid/cadence/observed_leaf_seeds.esm` writes its state
        # equation as an `aggregate` whose body is `D(index(u,i))`. A binding
        # that only recognises a bare `D(u)` LHS misses the state entirely and
        # then reports every unknown as algebraic.
        path = joinpath(TESTUTILS_REPO_ROOT, "tests", "valid", "cadence",
                        "observed_leaf_seeds.esm")
        if _require_fixture(path)
            m = EarthSciAST.load(path).models["ObservedLeafSeeds"]
            @test ode_states(m) == ["u"]
            @test observed_unknowns(m) == ["geom", "geom_chain", "k_scaled", "u_scaled"]
            @test algebraic_unknowns(m) == String[]
            @test discrete_parameters(m) == ["Kdiff"]
            @test constant_parameters(m) == ["dx"]
        end
    end

    @testset "observed_definitions resolves out of declaration order" begin
        path = joinpath(_CLS_DIR, "fixtures", "observed_chain.esm")
        m = EarthSciAST.load(path).models["M"]
        defs = observed_definitions(m)
        @test sort!(collect(keys(defs))) == ["y", "z"]
        # `z ~ y*2` is declared FIRST and resolved LAST; the definition it maps
        # to is its own equation's RHS, not the first equation in the file.
        @test defs["z"] isa OpExpr
        @test observed_definition(m, "y") === defs["y"]
        @test observed_definition(m, "x") === nothing   # an ODE state, not observed
    end

    @testset "system_kind derivation: order is sde → pde → nonlinear → ode" begin
        pde = EarthSciAST.load(joinpath(_CLS_DIR, "fixtures", "system_kind_pde.esm"))
        # A spatial `D` (wrt an axis) and the grad/div/laplacian sugar are both
        # the pde signal; neither spelling is canonical.
        @test system_kind(pde.models["Transient"]) == "pde"
        @test system_kind(pde.models["SugarOps"]) == "pde"
        # pde BEFORE nonlinear: a steady-state PDE has no time derivative.
        @test system_kind(pde.models["SteadyState"]) == "pde"
        @test has_time_derivative(pde.models["SteadyState"]) == false
        # sde BEFORE pde: there is no SPDESystem to select.
        @test system_kind(pde.models["StochasticSpatial"]) == "sde"
        @test has_spatial_derivative(pde.models["StochasticSpatial"]) == true

        # The document declares ONE domain shared by all four models, which is
        # the discriminator: a binding deriving `pde` from the presence of a
        # domain block classifies all four alike.
        @test pde.domain !== nothing
    end

    @testset "system_kind_mismatch fires only on a contradicting field" begin
        variants = EarthSciAST.load(
            joinpath(_CLS_DIR, "fixtures", "system_kind_variants.esm"))
        # `Declared` carries a field that AGREES — no mismatch.
        @test declared_system_kind_mismatch(variants.models["Declared"]) === nothing
        # An absent field is never a mismatch.
        @test declared_system_kind_mismatch(variants.models["Sde"]) === nothing

        # A contradicting field is reported as (declared, derived).
        sde = variants.models["Sde"]
        wrong = Model(sde.variables, sde.equations; system_kind="ode")
        @test declared_system_kind_mismatch(wrong) == ("ode", "sde")
        errs = validate_structural(EsmFile("1.0.0", Metadata("M");
                                           models=Dict("Wrong" => wrong)))
        @test any(e -> e.error_type == "system_kind_mismatch", errs)
    end

    @testset "a distribution alone is NOT Brownian; an update alone is not either" begin
        m = EarthSciAST.load(
            joinpath(_CLS_DIR, "fixtures", "parameter_cadences.esm")).models["M"]
        # The two easy mistakes the fixture is built to catch.
        @test !("p_sampled" in brownian_parameters(m))
        @test !("p_uniform_sampled" in brownian_parameters(m))
        @test brownian_parameters(m) == ["p_wiener"]
        # The handler and `from` value forms classify exactly like `expression`.
        @test "p_handler" in discrete_parameters(m)
        @test "p_data" in discrete_parameters(m)
    end
end

File: pkg/earthsci-ast-ts/src/classification.test.ts
/**
 * Cross-language conformance for the esm 1.0.0 classification API
 * (esm-spec §6.3.1), driven by the shared oracle in
 * `tests/conformance/classification/`.
 *
 * The manifest lists TypeScript under `bindings_required`, so every fixture
 * there must pass here. The goldens are authored, not produced by any binding,
 * so this is a comparison against the spec rather than against another
 * implementation.
 */

import { describe, it, expect } from 'vitest'
import { readFileSync } from 'fs'
import { join } from 'path'
import { loadString } from './parse.js'
import {
  odeStates,
  observedUnknowns,
  algebraicUnknowns,
  isOdeState,
  brownianParameters,
  discreteParameters,
  sampledParameters,
  constantParameters,
  systemKind,
  declaredSystemKind,
  effectiveSystemKind,
  classifyDocument,
  unknowns,
  parameters,
} from './classification.js'
import { leafCadence } from './cadence.js'
import { flatten } from './flatten.js'
import type { EsmFile, Model } from './types.js'
import { fixturesDir } from './test-helpers.js'

const classificationDir = join(fixturesDir(), 'conformance', 'classification')

// The INDEXED LHS spelling of an arrayed definition (esm-spec §6.3.1), which the
// `classification` category cannot state: it carries no `index` or `aggregate`
// LHS at all, which is why four of the five bindings drifted onto the same wrong
// answer independently. Same golden shape, so the same driver reads it.
// See tests/conformance/classification_indexed_lhs/README.md.
const indexedLhsDir = join(fixturesDir(), 'conformance', 'classification_indexed_lhs')

interface GoldenEntry {
  ode_states: string[]
  observed_unknowns: string[]
  algebraic_unknowns: string[]
  brownian_parameters: string[]
  discrete_parameters: string[]
  sampled_parameters: string[]
  constant_parameters: string[]
  system_kind: string
  declared_system_kind?: string | null
}

/**
 * esm-libraries-spec §4.7.5 step 4's flattened buckets, where a golden pins
 * them. Optional: only the categories whose fixtures make the distinction
 * interesting carry it.
 */
interface GoldenFlattenBuckets {
  state_variables: string[]
  observed_variables: string[]
  algebraic_variables: string[]
}

interface Manifest {
  bindings_required: string[]
  fixtures: { id: string; fixture: string; golden: string; pins: string }[]
}

function runClassificationCategory(label: string, dir: string): void {
  const manifest: Manifest = JSON.parse(readFileSync(join(dir, 'manifest.json'), 'utf-8'))

  describe(label, () => {
    it('names typescript as a required binding', () => {
      expect(manifest.bindings_required).toContain('typescript')
    })

    for (const entry of manifest.fixtures) {
      describe(entry.id, () => {
        const doc = loadString(readFileSync(join(dir, entry.fixture), 'utf-8')) as {
          models: { [k: string]: unknown }
        }
        const golden: {
          models: { [k: string]: GoldenEntry }
          flatten_buckets?: GoldenFlattenBuckets
        } = JSON.parse(readFileSync(join(dir, entry.golden), 'utf-8'))
        const actual = classifyDocument(doc.models)

        const buckets = golden.flatten_buckets
        if (buckets !== undefined) {
          // Which flattened MAP an unknown lands in is a different question
          // from which §6.3.1 set classifies it, and §4.7.5 step 4 answers it
          // with two maps that are NOT a partition: an arrayed observed is in
          // `observedVariables` (an equation defines it) AND in
          // `stateVariables` (it materializes into a buffer the solver
          // allocates). Issue #270.
          it('files the unknowns in the §4.7.5 step-4 buckets, dual membership and all', () => {
            const flat = flatten(doc as unknown as EsmFile)
            expect(Object.keys(flat.stateVariables)).toEqual(buckets.state_variables)
            expect(Object.keys(flat.observedVariables)).toEqual(buckets.observed_variables)
            expect(Object.keys(flat.algebraicVariables)).toEqual(buckets.algebraic_variables)
          })
        }

        it(`classifies exactly the model nodes the golden names (${entry.pins.slice(0, 60)}…)`, () => {
          expect(Object.keys(actual).sort()).toEqual(Object.keys(golden.models).sort())
        })

        for (const [path, expected] of Object.entries(golden.models)) {
          describe(path, () => {
            it('partitions the unknowns as the golden says', () => {
              expect(actual[path].odeStates).toEqual(expected.ode_states)
              expect(actual[path].observedUnknowns).toEqual(expected.observed_unknowns)
              expect(actual[path].algebraicUnknowns).toEqual(expected.algebraic_unknowns)
            })

            it('partitions the parameters as the golden says', () => {
              expect(actual[path].brownianParameters).toEqual(expected.brownian_parameters)
              expect(actual[path].discreteParameters).toEqual(expected.discrete_parameters)
              expect(actual[path].sampledParameters).toEqual(expected.sampled_parameters)
              expect(actual[path].constantParameters).toEqual(expected.constant_parameters)
            })

            it('derives the system kind', () => {
              expect(actual[path].systemKind).toBe(expected.system_kind)
            })

            if (Object.prototype.hasOwnProperty.call(expected, 'declared_system_kind')) {
              it('reports the declared system kind verbatim', () => {
                expect(actual[path].declaredSystemKind).toBe(expected.declared_system_kind ?? null)
              })
            }

            it('THE UNKNOWN SETS PARTITION: disjoint, and together the unknowns', () => {
              const model = modelAt(doc.models, path)
              const parts = [
                actual[path].odeStates,
                actual[path].observedUnknowns,
                actual[path].algebraicUnknowns,
              ]
              const union = parts.flat()
              expect([...union].sort()).toEqual(unknowns(model))
              // Disjoint: no name appears in two of the three.
              expect(new Set(union).size).toBe(union.length)
            })

            it('THE PARAMETER SETS PARTITION: disjoint, and together the parameters', () => {
              const model = modelAt(doc.models, path)
              const parts = [
                actual[path].brownianParameters,
                actual[path].discreteParameters,
                actual[path].sampledParameters,
                actual[path].constantParameters,
              ]
              const union = parts.flat()
              expect([...union].sort()).toEqual(parameters(model))
              expect(new Set(union).size).toBe(union.length)
            })

            it('isOdeState agrees with odeStates, for every declared unknown', () => {
              const model = modelAt(doc.models, path)
              for (const name of unknowns(model)) {
                expect(isOdeState(model, name)).toBe(expected.ode_states.includes(name))
              }
            })

            it('the standalone accessors agree with classifyDocument', () => {
              const model = modelAt(doc.models, path)
              expect(odeStates(model)).toEqual(expected.ode_states)
              expect(observedUnknowns(model)).toEqual(expected.observed_unknowns)
              expect(algebraicUnknowns(model)).toEqual(expected.algebraic_unknowns)
              expect(brownianParameters(model)).toEqual(expected.brownian_parameters)
              expect(discreteParameters(model)).toEqual(expected.discrete_parameters)
              expect(sampledParameters(model)).toEqual(expected.sampled_parameters)
              expect(constantParameters(model)).toEqual(expected.constant_parameters)
              expect(systemKind(model)).toBe(expected.system_kind)
              // §8 item 11: `declared ?? derived`, never null.
              expect(effectiveSystemKind(model)).toBe(
                declaredSystemKind(model) ?? expected.system_kind,
              )
            })
          })
        }
      })
    }
  })
}

runClassificationCategory('classification conformance (esm-spec §6.3.1)', classificationDir)
runClassificationCategory(
  'classification conformance: the INDEXED LHS spelling (esm-spec §6.3.1)',
  indexedLhsDir,
)

/** Resolve a dot-path like `Parent.Child` to its model node. */
function modelAt(models: { [k: string]: unknown }, path: string): Model {
  const [head, ...rest] = path.split('.')
  let node = models[head] as Model
  for (const part of rest) {
    node = (node.subsystems as { [k: string]: Model })[part]
  }
  return node
}

describe('effectiveSystemKind (API_SPEC.md §8 item 11)', () => {
  const derivedOde: Model = {
    variables: { x: { type: 'unknown', default: 0 } },
    equations: [{ lhs: { op: 'D', args: ['x', 't'] }, rhs: 1 }],
  } as unknown as Model

  it('falls back to the DERIVED kind when nothing is declared', () => {
    expect(declaredSystemKind(derivedOde)).toBeNull()
    expect(systemKind(derivedOde)).toBe('ode')
    expect(effectiveSystemKind(derivedOde)).toBe('ode')
  })

  it('prefers the DECLARED kind when the document states one', () => {
    const declared = { ...derivedOde, system_kind: 'nonlinear' } as unknown as Model
    expect(declaredSystemKind(declared)).toBe('nonlinear')
    expect(systemKind(declared)).toBe('ode')
    expect(effectiveSystemKind(declared)).toBe('nonlinear')
  })
})

/**
 * esm-spec §6.3.1 admits TWO LHS spellings for the equation that DEFINES an
 * unknown — bare (`y ~ f(…)`) and indexed (`y[i] ~ f(…)`, which defines the
 * whole array `y`) — and neither is restricted by rank: "the defining form is
 * read through the LHS's BASE NAME … so an arrayed definition is observed
 * exactly as its scalar counterpart is".
 *
 * Both spellings of the indexed one must credit `w` an OBSERVED: the bare
 * `index(w, i)` LHS that §6.3.1's worked example names verbatim, and the
 * `aggregate{k}(index(w, k))` shell that documents in this repo actually use
 * (both are written out in `tests/conformance/classification_indexed_lhs/`).
 * `baseVariableName` stopped at the aggregate shell, so
 * the second spelling was credited to nobody and {@link algebraicUnknowns}
 * claimed `w` by elimination.
 *
 * That is not bookkeeping. §6.3.1: `algebraic_unknowns` seeds the CONTINUOUS
 * cadence partition, whereas an observed's cadence resolves through its
 * defining RHS — "so misclassifying an arrayed definition as algebraic pushes
 * build-time work … onto the per-timestep hot path". The partition assertion is
 * what pins that consequence.
 *
 * The classification conformance corpus above cannot catch this: none of its
 * fixtures carries an `index` or `aggregate` LHS at all.
 *
 * See issue #232 (Julia), issue #231 / PR #237 (Python), PR #250.
 */
describe('an ARRAYED definition is observed, in every LHS spelling (§6.3.1)', () => {
  const aggregateIndexed = {
    op: 'aggregate',
    args: [],
    output_idx: ['k'],
    ranges: { k: { from: 'lev' } },
    expr: { op: 'index', args: ['w', 'k'] },
  }
  const spellings: { [label: string]: unknown } = {
    bare: 'w',
    indexed: { op: 'index', args: ['w', 'k'] },
    'aggregate of indexed': aggregateIndexed,
  }

  for (const [label, lhs] of Object.entries(spellings)) {
    describe(label, () => {
      const model = {
        variables: {
          u: { type: 'unknown', units: '1' },
          w: { type: 'unknown', units: '1', shape: ['lev'] },
        },
        equations: [
          { lhs: { op: 'D', args: ['u'], wrt: 't' }, rhs: 'w' },
          { lhs, rhs: 2 },
        ],
      } as unknown as Model

      it('puts `w` in observedUnknowns', () => {
        expect(observedUnknowns(model)).toEqual(['w'])
      })

      // The three sets PARTITION, so crediting the definition must also take
      // `w` out of the algebraic bucket, not merely add it to the observed one.
      it('leaves algebraicUnknowns empty', () => {
        expect(algebraicUnknowns(model)).toEqual([])
      })

      it('leaves the ODE partition untouched', () => {
        expect(odeStates(model)).toEqual(['u'])
      })

      // THE CONSEQUENCE, not just the bookkeeping. `w`'s definition is
      // state-free, so an observed seeds from its RHS's class and folds at bind.
      // While the aggregate spelling was mis-credited to `algebraicUnknowns` it
      // seeded CONTINUOUS instead, putting build-time work on the per-timestep
      // hot path — exactly what §6.3.1 warns this misclassification costs.
      it('folds a state-free arrayed observed at bind', () => {
        expect(leafCadence(model, 'w')).toBe('const')
      })
    })
  }

  // The aggregate unwrap recognises an `index` body ONLY. An `aggregate` whose
  // body is a `D` is the whole-array spelling of a TENDENCY — it makes an ODE
  // state, and crediting it as a definition would hand a state's derivative RHS
  // to the units and cadence passes as if it defined the state.
  it('does NOT credit an aggregate over a derivative as a definition', () => {
    const model = {
      variables: { u: { type: 'unknown', units: '1', shape: ['lev'] } },
      equations: [
        {
          lhs: {
            op: 'aggregate',
            args: [],
            output_idx: ['k'],
            ranges: { k: { from: 'lev' } },
            expr: { op: 'D', args: [{ op: 'index', args: ['u', 'k'] }], wrt: 't' },
          },
          rhs: 1,
        },
      ],
    } as unknown as Model
    expect(odeStates(model)).toEqual(['u'])
    expect(observedUnknowns(model)).toEqual([])
    expect(algebraicUnknowns(model)).toEqual([])
  })
})
